### ISSUE TITLE
Restart/Shutdown of qtile from extension.CommandSet not working

### ISSUE THREAD

**Issue by Tokariew** [qtile/qtile#2730]
# Issue description

Tried to use extensions.CommandSet to write simple "Power menu" 

After selecting restart qtile or logout, qtile freezes for couple of seconds.

# Qtile version

0.18.0

# Stack traces

* `~/.xsession-errors`

```
Traceback (most recent call last):
  File "/usr/lib64/python3.9/asyncio/streams.py", line 677, in read
    block = await self.read(self._limit)
  File "/usr/lib64/python3.9/asyncio/streams.py", line 684, in read
    await self._wait_for_data('read')
  File "/usr/lib64/python3.9/asyncio/streams.py", line 517, in _wait_for_data
    await self._waiter
asyncio.exceptions.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib64/python3.9/asyncio/tasks.py", line 492, in wait_for
    fut.result()
asyncio.exceptions.CancelledError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/tokariew/.local/lib/python3.9/site-packages/libqtile/ipc.py", line 183, in async_send
    read_data = await asyncio.wait_for(reader.read(), timeout=10)
  File "/usr/lib64/python3.9/asyncio/tasks.py", line 494, in wait_for
    raise exceptions.TimeoutError() from exc
asyncio.exceptions.TimeoutError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/tokariew/.local/bin/qtile", line 8, in <module>
    sys.exit(main())
  File "/home/tokariew/.local/lib/python3.9/site-packages/libqtile/scripts/main.py", line 55, in main
    options.func(options)
  File "/home/tokariew/.local/lib/python3.9/site-packages/libqtile/scripts/cmd_obj.py", line 169, in cmd_obj
    ret = run_function(obj, args.function, args.args)
  File "/home/tokariew/.local/lib/python3.9/site-packages/libqtile/scripts/cmd_obj.py", line 131, in run_function
    ret = client.call(funcname, *args)
  File "/home/tokariew/.local/lib/python3.9/site-packages/libqtile/command/client.py", line 118, in call
    if name not in self.commands:
  File "/home/tokariew/.local/lib/python3.9/site-packages/libqtile/command/client.py", line 138, in commands
    return self._command.execute(command_call, (), {})
  File "/home/tokariew/.local/lib/python3.9/site-packages/libqtile/command/interface.py", line 231, in execute
    status, result = self._client.send((
  File "/home/tokariew/.local/lib/python3.9/site-packages/libqtile/ipc.py", line 163, in send
    return asyncio.run(self.async_send(msg))
  File "/usr/lib64/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/lib64/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/home/tokariew/.local/lib/python3.9/site-packages/libqtile/ipc.py", line 185, in async_send
    raise IPCError("Server not responding")
libqtile.ipc.IPCError: Server not responding
2021-08-30 19:13:20,448 ERROR libqtile loop.py:_handle_exception():L59 [Errno 32] Broken pipe
NoneType: None
```

* `~/.local/share/qtile/qtile.log`

Log file have only last 2 lines from .xsession-errors

# Configuration

```python
from libqtile import extension

key_exensions = Key([mod], 's', lazy.run_extension(extension.CommandSet(
    commands={
        'reboot': 'reboot',
        'shutdown': 'shutdown',
        'suspend': 'systemctl suspend',
        'lock session': '/home/tokariew/.local/bin/lockme',
        'restart qtile': 'qtile cmd-obj -o cmd -f restart',
        'logout': 'qtile cmd-obj -o cmd -f shutdown',
        },
    dmenu_lines=6))),

keys.extend(key_exensions)
```

**Comment by m-col** [comment]
Does it work if you just call those commands directly i.e. from a terminal?

**Comment by elParaguayo** [comment]
Dammit @m-col, I was just posting that exact same question and yours popped up as I was about to hit "Comment".

**Comment by Tokariew** [comment]
Yes, `qtile cmd-obj -o cmd -f restart` and `qtile cmd-obj -o cmd -f shutdown` work fine from terminal

**Comment by elParaguayo** [comment]
OK - thanks for confirming. I'll take a look and see if I can recreate and, if so, try to fix it!

**Comment by elParaguayo** [comment]
Yup. I get broken pipe errors too. Will investigate.

**Comment by elParaguayo** [comment]
I can also recreate the issue if I bind a key to:
```python
lazy.function(lambda _: os.system("qtile cmd-obj -o cmd -f restart")
```
So the issue seems to be killing qtile in the middle of a `lazy` call...

**Comment by m-col** [comment]
I suppose this isn't too surprising - the server can't continue its IPC if it's decided to shut down. Perhaps we can run the shutdown/restart commands as a new asyncio task so the current IPC can complete before starting the shutdown/restart.

**Comment by elParaguayo** [comment]
But why doesn't `lazy.restart()` give the same problem?

**Comment by m-col** [comment]
I guess it'd be part of the IPC that is causing the issue, and there is no IPC when run that way. Maybe?

**Comment by elParaguayo** [comment]
It's weird because the extension is run via a `lazy` call too.

Could take a bit of time to get to the bottom of this one.

**Comment by m-col** [comment]
It is run via `lazy` but not via IPC. Reading through the code it seems as though it should all be fine though, as "restarting" involves just setting a flag and stopping the event loop, so the current task should finish successfully, along with the IPC. Perhaps there is another task scheduled at some point in the IPC code that I'm missing. The shutdown over IPC should should work but the IPC should "close unexpectedly" as the server shuts down the core as part of that command.

**Comment by elParaguayo** [comment]
That's my point (and I've probably missed something) - I don't think either `lazy.restart` or `lazy.run_extension` are using IPC. The difference seems to be that the extension calls `os.system` which then starts an IPC call. It seems killing the IPC from within `lazy` is a problem as running `qtile cmd-obj -o cmd -f restart` from a shell is fine.

**Comment by Tokariew** [comment]
Similar situation occur when trying to run `qtile run-cmd`, from `lazy.run_extension` or `lazy.function`.

**Comment by elParaguayo** [comment]
This is where the lazy call is executed:
https://github.com/qtile/qtile/blob/673b40d945dd166b37903ad3032fd72c9aea7db3/libqtile/command/interface.py#L312-L317

Line 313.

This is called from `manager.py` when it processes the key presses:
https://github.com/qtile/qtile/blob/057918dc2ef3ff830550255d7ebc7d5b76d31811/libqtile/core/manager.py#L355-L361

**Comment by zordsdavini** [comment]
Just pulled dev version (0.18.2.dev130+g3fe5e0d1) and all CommandSet keys are not working. No log. For dmenu extension it looks working, except dmenu_run doesn't take theme

**Comment by elParaguayo** [comment]
Can you give me an example from your config that I can test?

**Comment by zordsdavini** [comment]
This should work with ExKey (if need can rewrite on Key)

```
    EzKey("M-C-q", lazy.run_extension(extension.CommandSet(
        commands={
            'a': 'xterm'
            },
        foreground='#FF0000', selected_background='#000000')),
        desc='test'),

```

**Comment by elParaguayo** [comment]
The only thing that changed with extensions between 0.18.1 and latest as some error handling for colours so I don't think it's that.

Have you also upgraded python recently? I wonder if something's changed with Popen.communicate. Dmenu doesn't seem to be opening and just returns an empty string.

Will keep digging.

**Comment by zordsdavini** [comment]
Run through commits and tested on xephyr. The 4c0988c2e155ff33b0179d22db653454635be814 commit broke that. The log writes `cannot grab keyboard`

**Comment by zordsdavini** [comment]
## Color test on dmenu_run

Config (other dmenu extensions mostly have different colors, so something from defaults)

```
    FONT = 'Hack Nerd Font'
    FONT_SIZE = 14
    color_data = {
            'colors': {
                "color0": "#32302f",
                "color1": "#fb4934",
                "color2": "#b8bb26",
                "color3": "#fabd2f",
                "color4": "#83a598",
                "color5": "#d3869b",
                "color6": "#8ec07c",
                "color7": "#d5c4a1",
                "color8": "#665c54",
                "color9": "#fb4934",
                "color10": "#b8bb26",
                "color11": "#fabd2f",
                "color12": "#83a598",
                "color13": "#d3869b",
                "color14": "#8ec07c",
                "color15": "#fbf1c7"
                }
            }

BLACK = "#1A1C1D"
GREEN = color_data['colors']['color2']
    ...
    EzKey("M-r", lazy.run_extension(extension.DmenuRun()), desc='dmenu run'),
    ...
extension_defaults = dict(
    dmenu_font=FONT + '-10',
    background=BLACK,
    foreground=GREEN,
    selected_background=GREEN,
    selected_foreground=BLACK,
    dmenu_height=24,
)

```
On broken commit d7016e1dc27680437d89efb322d5e7db03e6b8fe
![image](https://user-images.githubusercontent.com/4191548/145881506-486068eb-37cb-4b5b-b7d2-78f454863328.png)
On working
![image](https://user-images.githubusercontent.com/4191548/145881996-3cd39a14-7801-4941-b2ef-53e52cfc26d5.png)

**Comment by elParaguayo** [comment]
Hmm. Not sure why those colours aren't working. Need to look at that.

As for keyboard grabbing, need to ping @m-col about that one...

**Comment by zordsdavini** [comment]
@elParaguayo I can't understand why but only this line https://github.com/qtile/qtile/commit/d7016e1dc27680437d89efb322d5e7db03e6b8fe#diff-d6a1e31ba13325a250107626e8bccc693ae150cb58b7d428c7e1f398795b84fdR59 will broke the flow. In my case I don't pass any colors in keys configuration section - I want to use extension_default, so for color_check it has None. But again, it should continue but it does some magic

**Comment by elParaguayo** [comment]
Yeah, it should work. If the attribute has been set via a configuration parameter, it will retrieve the value. If not, it will return None. Either way that should work.

**Comment by elParaguayo** [comment]
If I revert 4c0988c then I don't get issue with the colours so I don't _think_ d7016e1 is part of the problem.

**Comment by elParaguayo** [comment]
Also, I need to change `CommandSet` to use `qtile.cmd_spawn` instead of `system` to execute programs...

**Comment by zordsdavini** [comment]
@elParaguayo this is two different bugs. The problem with colors are in DmenuRun. Thou I didn't test on other Dmenu variants as I set colors during extension configuration not by default values. Interesting part about qtile.cmd_spawn :)

**Comment by elParaguayo** [comment]
OK - I think we need to create two new issues then as neither of these are directly related to the original one.
- One for the extension not working after 4c0988c
- One for the colours in DmenuRun which probably is related to d7016e1

**Comment by zordsdavini** [comment]
@elParaguayo wanted to create two new issues, but github gives 400...

**Comment by elParaguayo** [comment]
Was that if you tried to use the template? If so, just create a blank issue instead: https://github.com/qtile/qtile/issues/new

**Comment by m-col** [comment]
If templates are still broken (sad face) let's just delete the template 
and then re-add it when they are working again

**Comment by elParaguayo** [comment]
@Tokariew is this issue still happening?

I just `qtile cmd-obj -o cmd -f shutdown` from a `CommandSet` extension and it worked with no hang and no error in the log.

**Comment by Tokariew** [comment]
@elParaguayo 
tested with 0.21 and it is no longer happening, seems to be fixed.
